### ISSUE TITLE
Plugin Manager overhaul

### ISSUE THREAD

**Issue by alexbruy** [qgis/qgis3.0_api#43]
While Plugin Manager was improved in QGIS 2.x it still missed some important functionality:

- it is not possible to easily install plugin from local plugin package (e.g. downloaded from author's site). Of course one can extract and copy file to the plugins directory but this is not very user friendly. Also users often confuses with two places: plugins directory inside QGIS tree (``<QGIS root>/python/plugins``) and user's plugins directory (``~/.qgis2/plugins``)
- there is no support for directory-based plugins repositories
- access to the Plugin Manager's UI is quite limited, e.g. it is not possible to start Plugin Manager with predefined filter string

It would be good to address this in QGIS 3.0

**Comment by nyalldawson** [comment]
@alexbruy is this something which requires an API break?

**Comment by alexbruy** [comment]
Not sure if API break required, maybe it can be implemented without breaking API just extending it.

**Comment by mach0** [comment]
But I guess for 3.0 API breaks would be allowed anyway - so it is a good chance to get the plugin manager into a perfect state and learn from what is missing now.

**Comment by borysiasty** [comment]
More ideas from @alexbruy:

  - Don't lock the interface by the progress window when installing a plugin. Annoying when installing many plugins one by one. 

- Authentication per single plugin: if *401 Unauthorised* returned, ask for credentials (the same on repository xml as well as zip file requests)

- An API for (un)installing plugins. With multiple plugins at once, without unwanted popups.

**Comment by alexbruy** [comment]
Another possible improvement — replace or merge Plugin Manager with new Resourse Sharing plugin developed during GSoC. This way we will have single platform for sharing all kinds of resources in QGIS

**Comment by pcav** [comment]
+1

**Comment by wonder-sk** [comment]
[copied to the issue description]

---
More ideas :-)
- fetch repository updates in background - avoid the modal dialog when user wants to open plugin manager
- do _not_ load all plugins on startup just to check if they are not broken - if I understand correctly this check is done with all plugins, including the disabled ones, so loading QGIS gets slow with increasing number of installed plugins even if they are disabled (please correct me if I am wrong)
- make the plugin info page nicer

**Comment by borysiasty** [comment]
@alexbruy what do you mean by directory-based plugins repositories?

**Comment by alexbruy** [comment]
@borysiasty I mean local repositories without web-server. E.g. directory with plugin packages and XML file (or even without it). This can be useful in situations when setting-up web-server is not possible, like small computer class or workshops when there is no internet access.

**Comment by borysiasty** [comment]
@alexbruy: what about just using the file:// protocol? It works properly at mine, on Debian and Windows. Or is it too inconvinient for end users?

**Comment by alexbruy** [comment]
Fine for me.

**Comment by NathanW2** [comment]
Does using file:// work currently in QGIS 2.x?

**Comment by borysiasty** [comment]
@NathanW2 yep, works for me in 2.14 (with file:///c:/blahblah and file:///c:\blahblah notation on Windows). It's just QNetworkRequest. What I didn't test (and probably doesn't work) is a local network. Let me know if you observe any problems.

**Comment by NathanW2** [comment]
Nice. That's a easy win

**Comment by rduivenvoorde** [comment]
Reading this thread, may I add another argument to either not even load any plugin, or load them whenever QGIS itself is loaded? 
I've had several requests from people here in NL which try to run QGIS in thin client environments, be it either citrix or vmware based stuff. But ALL see that QGIS (and then especially the python part) loads awfully slow (like in >2-4 minutes the first cold start). Some report it is a problem with virus scanners (which apparently have to check every *.pyc file then....) others don't know, but blame python (as even in their osgeo4w environments starting the python interpreter takes ten's of seconds.
If anybody has hints: please let me know.
BUT in either case: not preloading plugins would be a win for this setups.

**Comment by NathanW2** [comment]
That's a bit trigger because some plugins want to do stuff at load time. I
have a solution but can do it until 3.2

**Comment by NathanW2** [comment]
Tricky

**Comment by m-kuhn** [comment]
what should a plugin do if it's disabled?

**Comment by borysiasty** [comment]
Ok, so we can remove the preloading: https://github.com/qgis/QGIS/pull/5532

Do we want to merge it now and prepare a nicer error reporting for enabled plugins for 3.2?

@NathanW2 I'm curious what solution do you have? :) My idea is to make a better error dialog and store the changelog for later display in the manager, so the plugin is only marked as an invalid after a failed loading (and only for the session lifetime).

**Comment by borysiasty** [comment]
@wonder-sk @rduivenvoorde @alexbruy, how do you like the solution above?

**Comment by wonder-sk** [comment]
hi @borysiasty ! Do you mean  qgis/QGIS#5532 to avoid preloading of plugins? I like that!

**Comment by borysiasty** [comment]
@wonder-sk yes, just didn't want to disable an existing feature without common enthusiasm ;)
